lib/git_repo.rb

```diff
@@ -0,0 +1,40 @@
+# frozen_string_literal: true
+
+class GitRepo
+  attr_reader :path, :name
+
+  def initialize(path, name = nil)
+    @path = path
+    @name = name
+    @memoize = {}
+  end
+
+  def valid?
+    File.exist?("#{path}/.git")
+  end
+
+  def url
+    url = run("config --get remote.origin.url")
+    return if url.blank?
+
+    url.sub!(/\Agit@github\.com:/, "https://github.com/")
+    url.sub!(/\.git\z/, "")
+    url
+  end
+
+  def latest_local_commit
+    run "rev-parse HEAD"
+  end
+
+  protected
+
+  def run(cmd)
+    @memoize[cmd] ||= begin
+      return unless valid?
+      stdout, stderr, status = Open3.capture3("git #{cmd}", chdir: path)
+      status == 0 ? stdout.strip : nil
```

Comment:
I believe you can use [Discourse::Utils.execute_command](https://github.com/discourse/discourse/blob/c7a9da1f10848518046f696c17f16047e1cb276e/lib/discourse.rb#L24) which has certain protections like timeout built in.